// @fal-ai/serverless-proxy/src/express.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.route = void 0;
const tslib_1 = require("tslib");
const index_1 = require("./index");
/**
 * The default Express route for the fal.ai client proxy.
 */
exports.route = index_1.DEFAULT_PROXY_ROUTE;
/**
 * The Express route handler for the fal.ai client proxy.
 *
 * @param request The Express request object.
 * @param response The Express response object.
 * @param next The Express next function.
 */
const handler = (request, response, next) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    yield (0, index_1.handleRequest)({
        id: 'express',
        method: request.method,
        respondWith: (status, data) => response.status(status).json(data),
        getHeaders: () => request.headers,
        getHeader: (name) => request.headers[name],
        sendHeader: (name, value) => response.setHeader(name, value),
        getBody: () => tslib_1.__awaiter(void 0, void 0, void 0, function* () { return JSON.stringify(request.body); }),
    });
    next();
});
exports.handler = handler;
