// @fal-ai/serverless-proxy/src/svelte.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRequestHandler = void 0;
const tslib_1 = require("tslib");
const index_1 = require("./index");
/**
 * Creates the SvelteKit request handler for the fal.ai client proxy on App Router apps.
 * The passed credentials will be used to authenticate the request, if not provided the
 * environment variable `FAL_KEY` will be used.
 *
 * @param params the request handler parameters.
 * @returns the SvelteKit request handler.
 */
const createRequestHandler = ({ credentials, } = {}) => {
    const handler = ({ request }) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        const FAL_KEY = credentials || process.env.FAL_KEY || '';
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const responseHeaders = {
            'Content-Type': 'application/json',
        };
        return yield (0, index_1.handleRequest)({
            id: 'svelte-app-router',
            method: request.method,
            respondWith: (status, data) => new Response(JSON.stringify(data), {
                status,
                headers: responseHeaders,
            }),
            getHeaders: () => (0, index_1.fromHeaders)(request.headers),
            getHeader: (name) => request.headers.get(name),
            sendHeader: (name, value) => (responseHeaders[name] = value),
            getBody: () => tslib_1.__awaiter(void 0, void 0, void 0, function* () { return request.text(); }),
            resolveApiKey: () => Promise.resolve(FAL_KEY),
        });
    });
    return {
        requestHandler: handler,
        GET: handler,
        POST: handler,
    };
};
exports.createRequestHandler = createRequestHandler;
